crates/cargo-util-schemas/src/restricted_names.rs

```diff
@@ -1,54 +1,73 @@
 //! Helpers for validating and checking names like package and crate names.
 
-use anyhow::bail;
-use anyhow::Result;
+type Result<T> = std::result::Result<T, NameValidationError>;
+
+/// Error validating names in Cargo.
+#[derive(Debug, thiserror::Error)]
+pub enum NameValidationError {
+    #[error("{0} cannot be empty")]
+    Empty(&'static str),
+
+    #[error("invalid character `{ch}` in {what}: `{name}`, {reason}")]
+    InvalidCharacter {
+        ch: char,
+        what: &'static str,
+        name: String,
+        reason: &'static str,
+    },
+
+    #[error(
+        "profile name `{name}` is reserved\n{help}\n\
+         See https://doc.rust-lang.org/cargo/reference/profiles.html \
+         for more on configuring profiles."
+    )]
+    ProfileNameReservedKeyword { name: String, help: &'static str },
+
+    #[error("feature named `{0}` is not allowed to start with `dep:`")]
+    FeatureNameStartsWithDepColon(String),
+}
 
 /// Check the base requirements for a package name.
 ///
 /// This can be used for other things than package names, to enforce some
 /// level of sanity. Note that package names have other restrictions
 /// elsewhere. `cargo new` has a few restrictions, such as checking for
 /// reserved names. crates.io has even more restrictions.
-pub fn validate_package_name(name: &str, what: &str, help: &str) -> Result<()> {
+pub fn validate_package_name(name: &str, what: &'static str, help: &str) -> Result<()> {
```

Comment:
... I forgot to mark this as `pub(crate)`.  Could you add a commit doing that?  It made analyzing `NameValidationError` harder.